README.md
# TYPO3 Backend Module 'slug' v.3.0.0 beta
Official Repository of the 'slug' Backend Module for TYPO3 9.5 and 10.2.0

[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://paypal.me/typo3freelancer/5)

The Slug backend module is designed to help manage large amounts of slugs for pages and extension records. Currently, it provides a simple list for pages and custom records, which can be filtered with different parameters. Slugs can be edited and saved quickly and efficiently. The modules regenerate or save **all slugs of the current list view** with just one click. I have tested the functionality with 500 empty news records so far, without any problem.

## IMPORTANT NOTE!

The current version 3.0.0 is in BETA status and ONLY for TYPO3 version 10.2. or higher!
If you want to use it with TYPO3 9, please download an older version here:
https://github.com/koehlersimon/slug/releases/tag/v2.0.2

## INTRODUCTION

Please use the latest version from the official TYPO3 repository (https://extensions.typo3.org/extension/slug/), if you want to make sure that nothing happens to your website. In any case I highly recommend a database backup if you want to work in a live website!!! If you use the current version from github, you use it at your own risk!

## FEATURES

* NEW in v3.0.0: Show hidden and deleted records in the page list
* NEW in v3.0.0: Display up to 3000 records at once. But handle with care ;-)
* List only custom records from a specific page or folder, by using the "pid" parameter
* Add custom records of your extensions via TypoScript (see manual below)
* Quickly edit, save and regenerate slugs for pages and other record types (new in Version 2)
* Mass generation and storage of news slugs (up to 500 at once)
* List views filterable with different parameters
* Search engine entry Preview for pages, displays the updated slug as you type
* Uses TYPO3 core slug generation functions
* Extension configuration for default values like sorting, entries per page etc.

## CHANGELOG

See the changelog for more details:
https://github.com/koehlersimon/slug/blob/master/CHANGELOG.md

## USAGE

### Installation

* For Composer use ```composer require koehlersimon/slug```
* Download the latest version here: https://extensions.typo3.org/extension/slug/ or install it with the extension manager of your TYPO3 installation
* No further configuration is required, but you should delete all the backend caches after installation to make sure the extension is working properly.

## CUSTOM RECORDS

Note: To enable custom records functionality, you must activate the checkbox "Custom records enabled" in the extension settings panel!

Important to know: Editing the slugs works only if the desired table contains a field for the title and a field for the slug. The names of the fields can be determined by TypoScript. But be careful. If you use a wrong field, the slug extension can destroy your data. I take no responsibility for it. So it's best not to test in a live web site before.
Very important to know:
* If you want to use an image symbol, make sure the image exists. The slug extension is currently NOT checking this!
* You can only use tables that are correctly prepared for TYPO3 use
    * The configuration array **$GLOBALS['TCA']['tx_your_table_name']['columns']['your_slug_field']['config']** needs to exist in the TYPO3 system. Otherwise the system will throw errors.
    * The fields **crdate,tstamp,uid** AND your custom fields for the title and the slug need to exist in your table!

Here's the TypoScript code you will need to make a custom table work. Put it into the setup of your root page.
```typoscript
# Module configuration
module.tx_slug {
    settings{
        additionalTables{
            tx_news_domain_model_news{
                pid = 0
                label = News
                slugField = path_segment
                titleField = title
                icon = EXT:news/Resources/Public/Icons/news_domain_model_news.svg
            }
        }
    }
}
```

## KNOWN PROBLEMS

### Oops, an error occured!: Invalid configuration: "vendorName" is not set
This happens, when you try to use the extension with version 3.0.0 or higher, in a TYPO3 version 9.5 or lower version. Please make sure to use slug v2.0.2 when using TYPO3 version 9!

### Uncaught TYPO3 Exception: #1278450972: Class SIMONKOEHLER\Slug\Controller\PageController does not exist. Reflection failed.
This exception occurs after upgrading the extension from 2.0.0 to 2.0.xx because the most important namespaces in the extbase PHP code have changed. Therefore, it is essential to clear the system's main cache and autoload data.

### Slug generation for News records failed?

If a news record has no pid set in the database, the slug generation will fail. This may happen when you have imported news records from a third party extension or manually. Solution: Check, if all entries in the table 'tx_news_domain_model_news' have the field 'pid' set to a page or folder in the page tree.

### Error, when 'Unassigned site configurations' have been found

The error "Argument 2 passed to TYPO3\CMS\Core\Imaging\IconFactory::getIconForRecord() must be of the type array, null given..." can be a result of "Unassigned site configurations"

## Want to report an issue?

https://github.com/koehlersimon/slug/issues

## All other requests

**Contact:** https://simon-koehler.com/en/contact

## Do you want to learn TYPO3? Here's what you need:
Video Training TYPO3 9 LTS (German language)
https://www.digistore24.com/redir/246076/GOCHILLA/


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [3.0.0] - 2019-12-20
- Composer settings changed
- Backend interface design optimized
- New Feature: The pages list now shows hidden and even deleted pages. Those can not be edited yet, because the TYPO3 internal classes for generating slugs cause problems when pages are not in any tree root.
- New Feature: You can define a custom label field for page records in the extension setup

## [2.0.2] - 2019-12-20
- New Icon for the backend module
- Compatibility checked with TYPO3 10.2.0

## [2.0.1] - 2019-07-27
- Added new option to custom records: By setting "pid=" only records from the related parent page will be shown
- BUGFIX: Fixed bug in TCA fields, thanks to @Teddytrombone on Github
- Changed namespace from "GOCHILLA" to "SIMONKOEHLER", so please empty autoload cache after update from version 2.0.0
- Several code clean ups
- Bugfix: Solved title problems with custom records. Issue #15 on GitHub, thanks @alex-boehm
- Changed footer information and template

## [2.0.0] - 2019-07-08
- BREAKING: News module removed!
- NEW: Add as many custom tables as you want simply by TypoScript!

## [1.0.15] - 2019-07-07
- Added functionality to use ANY custom table from TypoScript settings, but this is only a BETA version for now!
- Added sorting by timestamp (page field "tstamp") for page slugs

## [1.0.14] - 2019-07-01
- Bugfix-Release
- Checked compatibility with TYPO3 9.5.8!

## [1.0.13] - 2019-05-14
- Bugfix-Release
- Due to a change in the basic functionality in relation to TYPO3 Ajax Requests, the extension was not compatible with TYPO3 9.5.6. This has now been fixed!

## [1.0.12] - 2019-01-09
- Bugfix-Release
- Now creates only unique slugs!
- Issue 'Duplicate entries for news' solved (https://github.com/koehlersimon/slug/issues/2)
- Several small code improvements and changes

## [1.0.11] - 2019-01-07

### Changed
- Code optimizations in List.html and NewsList.html template files

### Removed
- Standard doc header

### Added
- Added experimental but working action for tree views!
- Record count badge in table header of list views
- Bulk saving and generating Page slugs finally works!
- Table row and buttons for bulk editing of Page slugs


## [1.0.10] - 2019-01-05

### Changed
- Removed some files that are no longer in use
- General Optimization of the slug saving functions

### Removed
- Removed Ajax Route 'getPageInfo' and Action in 'AjaxController.php' because it is no longer needed at the moment
- Removed function 'googlePreview' from 'AjaxController.php'
- Removed function 'getNewsRecordInfo' from 'AjaxController.php'

### Added
- Possibiliy to disable the 'additional record info' button in Pages list in global Extension configuration
- Dynamic select menu options including translation of all labels into German and Spanish
- More stable check for the News module. First checks if module is active, then if News table exists
- Comparison of original News slug and freshly generated News slug
- Notification if the freshly generated slug is still the same like before
- Switch statement in List.html template to show different partials for record previews in the Pages list
- Language labels for record previews in the Pages list
- Section 'Known Problems' to the README.md file
- Input fields for news record slugs are now also highlighted when changed, analogous to the page records
- News icon for the button in the top right corner


## [1.0.9] - 2019-01-04

### Added
- Small bugfixes
- Basic documentation skeleton added, will be continued soon


## [1.0.8] - 2019-01-03

### Changed
- Changed extension settings type for 'newsMaxEntries' from type string to options
- Changed extension settings type for 'newsOrderBy' from type string to options
- Changed sorting of extension settings for 'News' tab
- Extended dropdown menu 'Entries per page' according to the values of the extension settings option select
- Replaced standard HTML selects of the filter form with Fluid ViewHelpers of type 'f:form.select'
- Several clean-ups of all PHP classes
- Cleaned up JavaScript codes, removed unnecessary lines of code
- Removed way too strict warnings when saving slugs
- Corrected validation and fixed some bugs in validation

### Added
- Dynamic page URL in Pages list added. When user changes the text of the slug input, the URL will change in realtime, too
- Bulk saving and generating functionality added to News view! (this will follow soon for pages, too)
- Added translations to buttons in Partials/GlobalHeader.html
- Added labels for select menus in filter, but still not implemented in the frontend
- Added sorting property 'is_siteroot' to the filter form for pages
- Added full page URL when page rootpage has a configured site, if not just shows the slug below the title


## [1.0.7] - 2019-01-01

### Changed
- Initial release of the beta version after some testing

### Added
- Created changelog file and let's go on from here, happy new year!


ext_tables.php
<?php
defined('TYPO3_MODE') || die('Access denied.');

call_user_func(
    function()
    {

        if (TYPO3_MODE === 'BE') {

            /***************
             * Make the extension configuration accessible
             */
            $extensionConfiguration = \TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance(
                \TYPO3\CMS\Core\Configuration\ExtensionConfiguration::class
            );
            $slugConfiguration = $extensionConfiguration->get('slug');

            print_r(\TYPO3\CMS\Core\Utility\ExtensionManagementUtility::getExtensionVersion('sysext'));

            /* Register main site module for slug editing
            *  NOTE: This function has to be replaced with the outcommented
            *  Function below!
            */
            \TYPO3\CMS\Extbase\Utility\ExtensionUtility::registerModule(
                'SIMONKOEHLER.Slug',
                'site',
                'slugs',
                '',
                [
                    'Page' => 'list,site,tree',
                    'Extension' => 'additionalTable',
                ],
                [
                    'access' => 'user,group',
                    'icon'   => 'EXT:slug/Resources/Public/Icons/slug-be-module.svg',
                    'labels' => 'LLL:EXT:slug/Resources/Private/Language/locallang_slugs.xlf',
                ]
            );

            /*
            * Prepared code for upcoming TYPO3 version 11!
            \TYPO3\CMS\Extbase\Utility\ExtensionUtility::registerModule(
                'Slug',
                'site',
                'slugs',
                '',
                [
                    \SIMONKOEHLER\Slug\Controller\PageController::class => 'list,site,tree',
                    \SIMONKOEHLER\Slug\Controller\ExtensionController::class => 'additionalTable'
                ],
                [
                    'access' => 'user,group',
                    'icon'   => 'EXT:slug/Resources/Public/Icons/slug-be-module.svg',
                    'labels' => 'LLL:EXT:slug/Resources/Private/Language/locallang_slugs.xlf',
                ]
            );
            */


        }

        \TYPO3\CMS\Core\Utility\ExtensionManagementUtility::addStaticFile('slug', 'Configuration/TypoScript', 'Slug');
    }
);


Classes/Controller/AjaxController.php
<?php
namespace SIMONKOEHLER\Slug\Controller;
use SIMONKOEHLER\Slug\Utility\HelperUtility;
use TYPO3\CMS\Core\Utility\GeneralUtility;
use TYPO3\CMS\Core\Database\ConnectionPool;
use TYPO3\CMS\Core\DataHandling\SlugHelper;
use TYPO3\CMS\Core\Exception\SiteNotFoundException;
use TYPO3\CMS\Backend\Utility\BackendUtility;
use TYPO3\CMS\Core\Http\HtmlResponse;
use TYPO3\CMS\Core\Http\JsonResponse;

/*
 * This file was created by Simon Köhler
 * https://simon-koehler.com
 */

class AjaxController extends \TYPO3\CMS\Extbase\Mvc\Controller\ActionController {

    /**
     * @var HelperUtility
     */
    protected $helper;

    /**
     * function savePageSlug
     *
     * @return void
     */
    public function savePageSlug(\Psr\Http\Message\ServerRequestInterface $request)
    {
        $queryParams = $request->getQueryParams();
        $queryBuilder = GeneralUtility::makeInstance(ConnectionPool::class)->getQueryBuilderForTable('pages');
        $this->helper = GeneralUtility::makeInstance(HelperUtility::class);
        $slug = $this->helper->returnUniqueSlug('page', $queryParams['slug'], $queryParams['uid'], 'pages', 'slug');
        $queryBuilder->getRestrictions()->removeAll();
        $statement = $queryBuilder
            ->update('pages')
            ->where(
                $queryBuilder->expr()->eq('uid', $queryBuilder->createNamedParameter($queryParams['uid'],\PDO::PARAM_INT))
            )
            ->set('slug',$slug) // Function "createNamedParameter" is NOT needed here!
            ->execute();

        if($statement){
            $responseInfo['status'] = '1';
            $responseInfo['slug'] = $slug;
        }
        else{
            $responseInfo['status'] = '0';
            $responseInfo['slug'] = $slug;
        }
        return new JsonResponse($responseInfo);
    }

    /**
     * @param ServerRequestInterface $request
     * @param ResponseInterface $response
     * @return ResponseInterface
     */
    public function saveRecordSlug(\Psr\Http\Message\ServerRequestInterface $request)
    {
        $queryParams = $request->getQueryParams();
        $uid = $queryParams['uid'];
        $table = $queryParams['table'];
        $slug = $queryParams['slug'];
        $slugField  = $queryParams['slugField'];
        $queryBuilder = GeneralUtility::makeInstance(ConnectionPool::class)->getQueryBuilderForTable($table);
        $queryBuilder->getRestrictions()->removeAll();
        $statement = $queryBuilder
            ->update($table)
            ->where(
                $queryBuilder->expr()->eq('uid', $queryBuilder->createNamedParameter($uid,\PDO::PARAM_INT))
            )
            ->set($slugField,$slug) // Function "createNamedParameter" is NOT needed here!
            ->execute();
        $responseInfo['status'] = $statement;
        $responseInfo['slug'] = $slug;
        return new JsonResponse($responseInfo);
    }

    /**
     * function slugExists
     *
     * @return void
     */
    public function slugExists(\Psr\Http\Message\ServerRequestInterface $request)
    {
        $queryParams = $request->getQueryParams();
        $queryBuilder = GeneralUtility::makeInstance(ConnectionPool::class)->getQueryBuilderForTable('pages');
        $queryBuilder->getRestrictions()->removeAll();
        $result = $queryBuilder
            ->count('slug')
            ->from('pages')
            ->where(
                $queryBuilder->expr()->eq('slug', $queryBuilder->createNamedParameter($queryParams['slug']))
            )
            ->execute()
            ->fetchColumn(0);
        return new HtmlResponse($result);
    }

    /**
     * function generatePageSlug
     *
     * @return void
     */
    public function generatePageSlug(\Psr\Http\Message\ServerRequestInterface $request)
    {
        $fieldConfig = $GLOBALS['TCA']['pages']['columns']['slug']['config'];
        $slugHelper = GeneralUtility::makeInstance(SlugHelper::class, 'pages', 'slug', $fieldConfig);
        $this->helper = GeneralUtility::makeInstance(HelperUtility::class);
        $queryParams = $request->getQueryParams();
        $queryBuilder = GeneralUtility::makeInstance(ConnectionPool::class)->getQueryBuilderForTable('pages');
        $queryBuilder->getRestrictions()->removeAll();
        $statement = $queryBuilder
            ->select('*')
            ->from('pages')
            ->where(
                $queryBuilder->expr()->eq('uid', $queryBuilder->createNamedParameter($queryParams['uid'],\PDO::PARAM_INT))
            )
            ->execute();
        while ($row = $statement->fetch()) {
            $slugGenerated = $slugHelper->generate($row, $row['pid']);
            break;
        }
        $slug = $this->helper->returnUniqueSlug('page', $slugGenerated, $row['uid'], 'pages', 'slug');
        $responseInfo['status'] = $statement;
        $responseInfo['slug'] = $slug;
        return new JsonResponse($responseInfo);
    }

    /**
     * function generateRecordSlug
     *
     * @return void
     */
    public function generateRecordSlug(\Psr\Http\Message\ServerRequestInterface $request)
    {
        $queryParams = $request->getQueryParams();
        $uid = $queryParams['uid'];
        $table = $queryParams['table'];
        $slugField  = $queryParams['slugField'];
        $titleField  = $queryParams['titleField'];

        $fieldConfig = $GLOBALS['TCA'][$table]['columns'][$slugField]['config'];
        $slugHelper = GeneralUtility::makeInstance(SlugHelper::class, $table, $slugField, $fieldConfig);
        $this->helper = GeneralUtility::makeInstance(HelperUtility::class);

        $queryBuilder = GeneralUtility::makeInstance(ConnectionPool::class)->getQueryBuilderForTable('pages');
        $queryBuilder->getRestrictions()->removeAll();
        $statement = $queryBuilder
            ->select('*')
            ->from($table)
            ->where(
                $queryBuilder->expr()->eq('uid', $queryBuilder->createNamedParameter($queryParams['uid'],\PDO::PARAM_INT))
            )
            ->execute();
        while ($row = $statement->fetch()) {
            $slugGenerated = $slugHelper->sanitize($row[$titleField]);
            break;
        }

        $responseInfo['status'] = $statement;
        $responseInfo['slug'] = $slugGenerated;
        return new JsonResponse($responseInfo);
    }

    /**
     * function loadTreeItemSlugs
     *
     * @return void
     */
    public function loadTreeItemSlugs(\Psr\Http\Message\ServerRequestInterface $request)
    {
        $queryParams = $request->getQueryParams();
        $this->helper = GeneralUtility::makeInstance(HelperUtility::class);
        $translations = $this->helper->getPageTranslationsByUid($queryParams['uid']);
        $root = BackendUtility::getRecord('pages',$queryParams['uid']);
        $languages = $this->helper->getLanguages();
        $html .= '<div class="well">';
        $html .= '<h2>'.$root['title'].' <small>'.$root['seo_title'].'</small></h2>';
        $html .= '<div class="input-group">'
                . '<span class="input-group-addon"><i class="fa fa-globe"></i></span>'
                . '<input type="text" data-uid="'.$root['uid'].'" value="'.$root['slug'].'" class="form-control slug-input page-'.$root['uid'].'">'
                . '<span class="input-group-btn"><button data-uid="'.$root['uid'].'" id="savePageSlug-'.$root['uid'].'" class="btn btn-default savePageSlug" title="Save slug"><i class="fa fa-save"></i></button></span>'
                . '</div>';
        foreach ($translations as $page) {
            foreach ($languages as $value) {
                if($value['uid'] === $page['sys_language_uid']){
                    $icon = $value['language_isocode'];
                }
            }
            $html .= '<h3>'.$page['title'].' <small>'.$page['seo_title'].'</small></h3>';
            $html .= '<div class="input-group">'
                . '<span class="input-group-addon">'.$icon.'</span>'
                . '<input type="text" data-uid="'.$page['uid'].'" value="'.$page['slug'].'" class="form-control slug-input page-'.$page['uid'].'">'
                . '<span class="input-group-btn"><button data-uid="'.$page['uid'].'" id="savePageSlug-'.$page['uid'].'" class="btn btn-default savePageSlug" title="Save slug"><i class="fa fa-save"></i></button></span>'
                . '</div>';
        }
        $html .= '</div>';
        return new HtmlResponse($html);
    }

}


Classes/Domain/Repository/PageRepository.php
<?php
namespace SIMONKOEHLER\Slug\Domain\Repository;
use TYPO3\CMS\Core\Utility\GeneralUtility;
use TYPO3\CMS\Core\DataHandling\SlugHelper;
use TYPO3\CMS\Core\Site\SiteFinder;
use TYPO3\CMS\Core\Exception\SiteNotFoundException;
use TYPO3\CMS\Core\Database\Query\Restriction\DeletedRestriction;

/*
 * This file was created by Simon Köhler
 * https://simon-koehler.com
 */

class PageRepository extends \TYPO3\CMS\Extbase\Persistence\Repository {

    protected $table = 'pages';
    protected $fieldName = 'slug';
    protected $languages;

    public function findAllFiltered($filterVariables) {
        $this->languages = $this->getLanguages();
        $queryBuilder = GeneralUtility::makeInstance(\TYPO3\CMS\Core\Database\ConnectionPool::class)->getQueryBuilderForTable('pages');
        $queryBuilder
           ->getRestrictions()
           ->removeAll();

        $query = $queryBuilder
            ->select('*')
            ->from('pages')
            ->orderBy($filterVariables['orderby'],$filterVariables['order']);

        switch ($filterVariables['status']) {

            // All pages
            case 'all':
                // simply add nothing to the query...
            break;

            // Only hidden
            case 'hidden':
                $query->andWhere(
                    $queryBuilder->expr()->eq('hidden', 1),
                    $queryBuilder->expr()->eq('deleted', 0)
                );
            break;

            // Only deleted
            case 'deleted':
                $query->where(
                    $queryBuilder->expr()->eq('deleted', 1)
                );
            break;

            // Only visible pages (default)
            default:
                $query->where(
                    $queryBuilder->expr()->eq('hidden', 0),
                    $queryBuilder->expr()->eq('deleted', 0)
                );
            break;
        }

        if($filterVariables['key']){
            $query->andWhere(
                $queryBuilder->expr()->like('slug',$queryBuilder->createNamedParameter('%' . $queryBuilder->escapeLikeWildcards($filterVariables['key']) . '%'))
            );
        }

        $sitefinder = \TYPO3\CMS\Core\Utility\GeneralUtility::makeInstance(SiteFinder::class);
        $statement = $query->execute();
        $output = array();

        while ($row = $statement->fetch()) {

            $row['flag'] = $this->getLanguageValue('flag',$row['sys_language_uid']);
            $row['isocode'] = $this->getLanguageValue('language_isocode',$row['sys_language_uid']);

            // If page is a translated page, set l10n_parent as PageUid
            if($row['l10n_parent'] > 0){
                $pageUid = $row['l10n_parent'];
            }
            else{
                $pageUid = $row['uid'];
            }

            // Try to get the Site configuration
            try {
                $site = $sitefinder->getSiteByPageId($pageUid);
                $siteConf = $site->getConfiguration();

                $row['site'] = $site;
                $row['hasSite'] = true;

                // Remove slash from base URL if neccessary
                if(substr($siteConf['base'], -1) === "/"){
                    $row['pageurl'] = substr($siteConf['base'], 0, -1);
                }
                else{
                    $row['pageurl'] = $siteConf['base'];
                }

                if($row['isocode']){
                    $row['pageurl'] = $row['pageurl'].'/'.$row['isocode'];
                }
            }
            catch (SiteNotFoundException $e) {
               $row['hasSite'] = false;
               $row['pageurl'] = '(N/A)';
            }

            array_push($output, $row);

        }
        return $output;
    }

    function getLanguageValue($field,$uid){
        foreach ($this->languages as $language) {
            if($language['uid'] === $uid){
                $output = $language[$field];
                break;
            }
            elseif($uid === 0){
                if($field === 'flag'){
                    $output = 'multiple';
                }
                else{
                    $output = '';
                }
                break;
            }
        }
        return $output;
    }

    public function getLanguages(){
        $queryBuilder = GeneralUtility::makeInstance(\TYPO3\CMS\Core\Database\ConnectionPool::class)->getQueryBuilderForTable('sys_language');
        $statement = $queryBuilder
            ->select('*')
            ->from('sys_language')
            ->execute();
        $output = array();
        while ($row = $statement->fetch()) {
            array_push($output, $row);
        }
        return $output;
    }

}


Classes/Controller/PageController.php
<?php
namespace SIMONKOEHLER\Slug\Controller;
use SIMONKOEHLER\Slug\Utility\HelperUtility;
use SIMONKOEHLER\Slug\Domain\Repository\PageRepository;
use TYPO3\CMS\Core\Imaging\Icon;
use TYPO3\CMS\Core\Imaging\IconFactory;
use TYPO3\CMS\Core\Utility\GeneralUtility;
use TYPO3\CMS\Core\Configuration\ExtensionConfiguration;
use TYPO3\CMS\Core\Site\SiteFinder;
use TYPO3\CMS\Backend\Tree\View\PageTreeView;
use TYPO3\CMS\Backend\Utility\BackendUtility;
use TYPO3\CMS\Core\Database\ConnectionPool;
use TYPO3\CMS\Core\Database\Query\Restriction\DeletedRestriction;

/*
 * This file was created by Simon Köhler
 * https://simon-koehler.com
 */

class PageController extends \TYPO3\CMS\Extbase\Mvc\Controller\ActionController {

    public $pageRepository;
    protected $iconFactory;
    protected $helper;
    protected $languages;
    protected $sites;
    protected $backendConfiguration;

    /**
    * @param PageRepository $pageRepository
    */
    public function __construct(PageRepository $pageRepository) {
         $this->pageRepository = $pageRepository;
         $this->iconFactory = GeneralUtility::makeInstance(IconFactory::class);
         $this->helper = GeneralUtility::makeInstance(HelperUtility::class);
         $this->backendConfiguration = GeneralUtility::makeInstance(ExtensionConfiguration::class)->get('slug');
         $this->sites = GeneralUtility::makeInstance(SiteFinder::class)->getAllSites();
    }

    /**
     * List all slugs from the pages table
     */
    protected function listAction()
    {

        // Check if filter variables are available, otherwise set default values from ExtensionConfiguration
        if($this->request->hasArgument('filter')){
            $filterVariables = $this->request->getArgument('filter');
        }
        else{
            $filterVariables['maxentries'] = $this->backendConfiguration['defaultMaxEntries'];
            $filterVariables['orderby'] = $this->backendConfiguration['defaultOrderBy'];
            $filterVariables['order'] = $this->backendConfiguration['defaultOrder'];
            $filterVariables['status'] = $this->backendConfiguration['defaultStatus'];
            $filterVariables['key'] = '';
        }

        // Set the order by options for fluid viewhelper f:form.switch
        $filterOptions['orderby'] = [
            ['value' => 'crdate', 'label' => $this->helper->getLangKey('filter.form.select.option.creation_date')],
            ['value' => 'tstamp', 'label' => $this->helper->getLangKey('filter.form.select.option.tstamp')],
            ['value' => 'title', 'label' => $this->helper->getLangKey('filter.form.select.option.title')],
            ['value' => 'slug', 'label' => $this->helper->getLangKey('filter.form.select.option.slug')],
            ['value' => 'sys_language_uid', 'label' => $this->helper->getLangKey('filter.form.select.option.sys_language_uid')],
            ['value' => 'is_siteroot', 'label' => $this->helper->getLangKey('filter.form.select.option.is_siteroot')],
            ['value' => 'doktype', 'label' => $this->helper->getLangKey('filter.form.select.option.doktype')]
        ];

        $filterOptions['order'] = [
            ['value' => 'DESC', 'label' => $this->helper->getLangKey('filter.form.select.option.descending')],
            ['value' => 'ASC', 'label' => $this->helper->getLangKey('filter.form.select.option.ascending')]
        ];

        $filterOptions['maxentries'] = [
            ['value' => '10', 'label' => '10'],
            ['value' => '20', 'label' => '20'],
            ['value' => '30', 'label' => '30'],
            ['value' => '40', 'label' => '40'],
            ['value' => '50', 'label' => '50'],
            ['value' => '60', 'label' => '60'],
            ['value' => '70', 'label' => '70'],
            ['value' => '80', 'label' => '80'],
            ['value' => '90', 'label' => '90'],
            ['value' => '100', 'label' => '100'],
            ['value' => '150', 'label' => '150'],
            ['value' => '200', 'label' => '200'],
            ['value' => '300', 'label' => '300'],
            ['value' => '400', 'label' => '400'],
            ['value' => '500', 'label' => '500'],
            ['value' => '1000', 'label' => '1000 (be careful!)'],
            ['value' => '1500', 'label' => '1500'],
            ['value' => '2000', 'label' => '2000'],
            ['value' => '3000', 'label' => '3000']
        ];

        $this->view->assignMultiple([
            'pages' => $this->pageRepository->findAllFiltered($filterVariables),
            'filter' => $filterVariables,
            'backendConfiguration' => $this->backendConfiguration,
            'beLanguage' => $GLOBALS['BE_USER']->user['lang'],
            'extEmconf' => $this->helper->getEmConfiguration('slug'),
            'filterOptions' => $filterOptions,
            'additionalTables' => $this->settings['additionalTables']
        ]);

    }


    /**
     * Generate a tree view
     */
    protected function treeAction()
    {

        if($this->request->hasArgument('siteRoot')){
            $siteRoot = $this->request->getArgument('siteRoot');
        }
        else{
            $siteRoot = $this->backendConfiguration['treeDefaultRoot'];
        }

        if(!$siteRoot || $siteRoot === 0){
            //Get the first existing site in the root level and its uid
            foreach ($this->sites as $site) {
                $siteRoot = $site->getRootPageId();
                break;
            }
        }

        if($this->request->hasArgument('depth')){
            $depth = $this->request->getArgument('depth');
        }
        else{
            $depth = $this->backendConfiguration['treeDefaultDepth'];
        }

        if($siteRoot){
            $args['siteRoot'] = $siteRoot;
            $args['depth'] = $depth;
            $siteRootRecord = BackendUtility::getRecord('pages',$siteRoot);
            $tree = GeneralUtility::makeInstance(PageTreeView::class);
            $tree->init('AND ' . $GLOBALS['BE_USER']->getPagePermsClause(1));
            $iconFactory = GeneralUtility::makeInstance(IconFactory::class);
            $icon = '<span class="page-icon">' . $iconFactory->getIconForRecord('pages', $siteRootRecord, Icon::SIZE_SMALL)->render() . '</span>';
            $tree->tree[] = array(
                'row' => $siteRootRecord,
                'icon' => $icon
            );
            $tree->getTree($siteRoot,$depth,'');
            $this->view->assignMultiple([
                'tree' => $tree->tree,
                'backendConfiguration' => $this->backendConfiguration,
                'extEmconf' => $this->helper->getEmConfiguration('slug'),
                'sites' => (array) $this->sites,
                'args' => $args
            ]);
        }
        else{
            $this->addFlashMessage('Error: No Site root found! PageController.php Line 130');
        }

    }

}
